crates/ruff_linter/src/settings/types.rs

```diff
@@ -635,6 +635,78 @@ impl Deref for CompiledPerFileIgnoreList {
     }
 }
 
+// This struct and its `new` implementation are adapted directly from `PerFileIgnore`, minus the
+// `negated` field
+#[derive(Debug, Clone)]
+pub struct PerFileVersion {
+    pub(crate) basename: String,
+    pub(crate) absolute: PathBuf,
+    pub(crate) version: ast::PythonVersion,
+}
+
+impl PerFileVersion {
+    pub fn new(pattern: String, version: ast::PythonVersion, project_root: Option<&Path>) -> Self {
+        let absolute = match project_root {
+            Some(project_root) => fs::normalize_path_to(&pattern, project_root),
+            None => fs::normalize_path(&pattern),
+        };
+
+        Self {
+            basename: pattern,
+            absolute,
+            version,
+        }
+    }
+}
+
+#[derive(Debug, Clone, CacheKey)]
+pub struct CompiledPerFileVersion {
+    pub absolute_matcher: GlobMatcher,
+    pub basename_matcher: GlobMatcher,
+    pub version: ast::PythonVersion,
+}
+
+#[derive(CacheKey, Clone, Debug, Default)]
+pub struct CompiledPerFileVersionList {
```

Comment:
Do we need to implement `Display` similar to `PerFileIgnores`? How does the output in `--show-settings` look like?